### ISSUE TITLE
Heapster stops retrieving CPU metrics when pod is under load

### ISSUE THREAD

**Issue by whites11** [kubernetes/heapster#1937]
Disclaimer: I can't be sure if this is heapster's fault or not, so please don't get angry if I am in the wrong spot.

My environment:

- kubernetes 1.8.4 on AWS, deployed with kops.
- Heapster 1.5.0 with influxdb sink and 60s metric resolution

When my pods are idle or low traffic everything's fine. My HPAs can get data out of heapster and I can see data on grafana, pulling it out from influxdb.

When I start loadtesting a pod (putting it under some traffic, starting at 10 rqs/second) i stop getting info about the CPU usage in grafana and HPAs start getting this:

```
Events:
  Type     Reason                        Age               From                       Message
  ----     ------                        ----              ----                       -------
  Warning  FailedGetResourceMetric       2m (x13 over 3h)  horizontal-pod-autoscaler  unable to get metrics for resource cpu: no metrics returned from heapster
  Warning  FailedComputeMetricsReplicas  2m (x13 over 3h)  horizontal-pod-autoscaler  failed to get cpu utilization: unable to get metrics for resource cpu: no metrics returned from heapster
```

After the load finishes, I get almost immediately back my CPU metrics both in influxdb and HPAs.
Please note that, during the same period, I never ever lose data about memory usage.

What I tried so far to solve the problem (and didn't work):

1) scale up heapster to two replicas
2) lower the metric resolution to various intervals, down to the minimum of 5s.

Any idea how to troubleshoot this?

**Comment by DirectXMan12** [comment]
Huh, fascinating.  Based on the initial symptoms, I'd guess something's being starved for CPU, probably either the kubelet or Heapster itself, ***but*** since you don't lose memory stats, that probably indicates a rate calculation issue (since CPU usage is actually a calculated rate metric).

That would imply that it's *probably* Kubelet that's starved for CPU, and collection is taking more than the Heapster metric collection interval.  This would in turn cause Heapster to get the same timestamps on two different pieces of data, meaning it would skip rate calculation.  This should show up in the Heapster logs if run with sufficiently high verbosity (`--v=10` is pretty much the highest we care about).

> lower the metric resolution to various intervals, down to the minimum of 5s.

Perhaps somewhat counterintuitively, this could make the problem worse if my theory is correct.  In that case, you'd actually want to *increase* the collection interval.

> Disclaimer: I can't be sure if this is heapster's fault or not, so please don't get angry if I am in the wrong spot.

I do try not to bite, I swear :-P

**Comment by whites11** [comment]
Thank you very much for getting back to me @DirectXMan12 

Your hypothesis make a lot of sense to me, I'll try increasing verbosity and post the logs.

> I do try not to bite, I swear :-P

Phewwww

**Comment by whites11** [comment]
So, a few updates on this case.
I raised verbosity to 10. It is so verbose that kubectl logs hangs up at a certaing point.
So i lowered verbosity to 5 and I discovered a difference in the log between a load situation or not a load situation. When CPU data is missing this is what I get in the log:

`Skipping rates for namespace:my-namespace/pod:my-pod-name-59587c855f-s9t99/container:app - different collection start time new:2018-01-25 09:52:13.317097259 +0000 UTC  old:2018-01-25 09:45:30.166309949 +0000 UTC`

This must be the reason why I get no CPU usage rate, but I have no clue on the cause.

**Comment by DirectXMan12** [comment]
so... that means that two different metrics from the same container have different start times listed.  Generally, it means that there's a bug in Kubelet.  The only time I've ever seen it like that before was when we had a prefixing bug in cAdvisor where it wasn't stripping off cgroup prefixes for systemd, but that was fixed a long time ago.  If you can manage to get the logs from level 10, you should see the exact responses received from the Kubelet, which might shed some light on the situation.

**Comment by whites11** [comment]
Hi,
First of all, sorry for late response.
I extracted a 1 minute log from the heapster container while the problem was occouring.
I tried hard to extract a smaller part of the log but I am having a very hard time understanding where a "scrape period" starts and where it ends. The problem is that the log file is 23MB.
I attached a gzipped version, hope it helps.

[all.log.gz](https://github.com/kubernetes/heapster/files/1685880/all.log.gz)
